marlowe-spec-test/src/Marlowe/Spec/Core/Semantics.hs

```diff
@@ -52,15 +121,260 @@ evalObservationTest interpret = reproducibleProperty "Eval Observation" do
+                ++ "Actual: " ++ show transactionOutput)
+        assert $ txOutEquals transactionOutput expected
+      _ -> fail "JSON parsing failed!"
+
+txOutEquals :: TransactionOutput -> TransactionOutput -> Bool
+txOutEquals
+  (TransactionOutput (TransactionOutputRecord_ext warnings1 payments1 (State_ext accounts1 choices1 boundValues1 minTime1 b1) contract1 a1))
+  (TransactionOutput (TransactionOutputRecord_ext warnings2 payments2 (State_ext accounts2 choices2 boundValues2 minTime2 b2) contract2 a2)) =
+    warnings1 == warnings2
+    && setEquals payments1 payments2
+    && setEquals (notZero accounts1) (notZero accounts2)
+    && setEquals choices1 choices2
+    && setEquals boundValues1 boundValues2
+    && minTime1 == minTime2
+    && contract1 == contract2
+    && a1 == a2
+    && b1 == b2
+  where
+    notZero = filter (\(_, i) -> less 0 i)
+txOutEquals a b = a == b
+
+setEquals :: Eq a => [a] -> [a] -> Bool
+setEquals l1 l2 =
+    all (flip elem l2) l1
+    && all (flip elem l1) l2
+
+-- TransactionBound.thy
+--
+-- lemma playTrace_only_accepts_maxTransactionsInitialState :
+--    "playTrace sl c l = TransactionOutput txOut ⟹
+--      length l ≤ maxTransactionsInitialState c"
+playTrace_only_accepts_maxTransactionsInitialStateTest :: InterpretJsonRequest -> TestTree
+playTrace_only_accepts_maxTransactionsInitialStateTest interpret = reproducibleProperty "Calling playTrace only accepts maxTransactionsInitialState"  do
+    contract <- run $ generateT $ genContract interpret
+    state <- run $ generateT $ genState interpret
+    transactions <- run $ generate $ arbitraryValidInputs state contract
+    let
+        req :: Request JSON.Value
+        req = PlayTrace (integer_of_int $ minTime state) contract transactions
+    RequestResponse res <- run $ liftIO $ interpret req
+
+    case JSON.fromJSON res of
+      JSON.Success (TransactionOutput TransactionOutputRecord_ext {}) -> do
+        monitor
+          ( counterexample $
+              "Request: " ++ showAsJson req ++ "\n"
+                ++ "Expected reponse to be quiescent" )
+        assert $ toInteger (length transactions) <= Arith.integer_of_nat (maxTransactionsInitialState contract)
+      JSON.Success _ -> pre False
+      _ -> fail "JSON parsing failed!"
+
+-- SingleInputTransactions.thy
+--
+-- theorem traceToSingleInputIsEquivalent:
+--    "playTrace sn co tral = playTrace sn co (traceListToSingleInput tral)"
+traceToSingleInputIsEquivalentTest :: InterpretJsonRequest -> TestTree
+traceToSingleInputIsEquivalentTest interpret = reproducibleProperty "Single input transactions"  do
+    (contract, State_ext _ _ _ startTime _, transactions) <- run $ generateT $ (do
+        c <- genContract interpret `suchThat` (/=Close) -- arbitraryValidInputs returns [] for the `Close` contract
+        s <- genState interpret
+        t <- liftGen $ arbitraryValidInputs s c
+        pure (c,s,t)) `suchThat` \(_,_,t) -> t /= traceListToSingleInput t
```

Comment:
This `suchThat` is what makes this test super slow (more than a minute in my machine).

Doing some manual tracing I noticed that most generated values returns the `[]` that you were trying to avoid.

![Screen Shot 2023-03-29 at 16 47 16](https://user-images.githubusercontent.com/2634059/228651094-defffaf4-7c6b-4526-b78f-5c0b1f7517a3.png)

I imagine that these are the results of Contracts without When's or with When's with empty cases (or just notify).

To avoid contract without When's, instead of discarding `Contract /= Close` I tried discarding using the following 
```
c <- genContract interpret `suchThat` (\c ->  Arith.integer_of_nat (maxTransactionsInitialState c) > 2)
```

but it didn't improve the situation.

Taking a closer look at the entries that passed the `suchThat`, I noticed that **all** of them were taken from the Escrow Contract. Which means that we are generating 90% random contract (that are discarded) and 10% golden contract (which don't have enough entropy)

Reply:
The distribution is bad, but there is also a portion of randomly generated tests that are taken into account (you can set the ratio of golden tests to 0 and it passes). On the one hand side we have to reduce the ratio of golden tests and on the other hand improve the generator for non-golden tests to produce less rejects

Reply:
- Reduced the ratio of golden contracts to 2% (like this every test run on average includes one escrow and one swap contract)
- Increased the weight for `When` contracts 

With the current setting I have almost always more random than golden contracts in the test set